README
Text::BSV::* version 1.03

DESCRIPTION

    This distribution includes the following modules:

        Text::BSV::BsvFileReader - Defines a class for reading BSV data
        from a file and then turning the data into an array of field names
        plus one hash encapsulating each record.

        Text::BSV::BsvListReader - Defines a class for reading BSV data
        from a list of strings containing the rows (without end-of-line
        characters), including the header row, and then turning the data
        into an array of field names plus one hash encapsulating each
        record.

        Text::BSV::BsvWriter - Defines a class for generating BSV data from
        a list of field names and a list of references to hashes that
        encapsulate records.  The class provides a method for returning the
        BSV data as a reference to an array of strings, and a method for
        writing the BSV data to a file.

        Text::BSV::BsvParsing - Contains helper functions for parsing and
        generating BSV data.  These functions are used by the other modules.
        Although there is no prohibition against client code using these
        functions directly, there is little likelihood of that being useful.

        Text::BSV::Exception - Defines a package for creating a
        Text::BSV::Exception object, which can be thrown using the Perl
        "die" function and caught using "eval".

    For information on the BSV (Bar-Separated Values) format itself, see the
    file named "bsv_format.txt".

REQUIREMENTS

    These modules require Perl 5, version 5.10.1 or later.

INSTALLATION

    To install the modules included in this distribution, simply copy the
    "Text" directory (with its BSV subdirectory and the files therein) to a
    location in @INC (for example, "/etc/perl" or "C:\Perl\site\lib").

AUTHOR

    Benjamin Fitch <blernflerkl@yahoo.com>

COPYRIGHT AND LICENSE

    Copyright 2010 by Benjamin Fitch

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.


Changes

Revision history for the Text::BSV distribution:

    Version    Date          Description
    -------    ----          -----------

    1.01       2010-09-11    Initial upload

    1.02       2010-09-13    Added Text::BSV::Exception

    1.03       2010-10-09    Fixed a documentation (POD) bug


lib/Text/BSV/BsvWriter.pm
####
# BsvWriter.pm:  A Perl module defining a class for generating BSV data from
# a list of field names and a list of references to hashes that encapsulate
# records.  The class provides a method for returning the BSV data as a
# reference to an array of strings, and a method for writing the BSV data to
# a file.
#
####
#
# Copyright 2010 by Benjamin Fitch.
#
# This library is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.
####
package Text::BSV::BsvWriter;

use 5.010001;
use strict;
use warnings;
use utf8;

use English "-no_match_vars";
use Hash::Util ("lock_keys");
use List::Util ("first", "max", "min", "sum");
use Scalar::Util ("looks_like_number");

use Text::BSV::BsvParsing;
use Text::BSV::Exception;

# Version:
our $VERSION = '1.03';

# Constants:
my $POUND     = "#";
my $SQ        = "'";
my $DQ        = "\"";
my $SEMICOLON = ";";
my $CR        = "\r";
my $LF        = "\n";
my $SPACE     = " ";
my $EMPTY     = "";
my $TRUE      = 1;
my $FALSE     = 0;

# Constructor:
sub new {
    my ($class, $field_names, $records) = @_;
    my %bsv_writer;

    unless (scalar(@{ $field_names }) >= 2) {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::ILLEGAL_ARGUMENT,
          "There must be at least two field names.");
    } # end unless

    unless (scalar(@{ $records }) >= 1) {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::ILLEGAL_ARGUMENT,
          "There must be at least one record.");
    } # end unless

    # Bless the hash into the class:
    bless \%bsv_writer, $class;

    # Restrict the hash keys:
    lock_keys(%bsv_writer, "_rows");

    # Generate the header row:
    push @{ $bsv_writer{"_rows"} }, generate_header_row($field_names);

    # Generate the records:
    for my $record (@{ $records }) {
        push @{ $bsv_writer{"_rows"} }, generate_row($record, $field_names);
    } # next $record

    # Return the object:
    return \%bsv_writer;
} # end constructor

# Methods:

sub get_rows {
    return $_[0]->{"_rows"};
} # end sub

sub write_to_file {
    my $bsv_writer = $_[0];
    my $file_path = $_[1];

    if (open my $BSV_FILE, ">:utf8", $file_path) {
        for my $row (@{ $bsv_writer->{"_rows"} }) {
            say $BSV_FILE $row;
        } # next $row

        close $BSV_FILE;
    }
    else {
        die Text::BSV::Exception->new($Text::BSV::Exception::IO_ERROR,
          "Couldn't open $DQ$file_path$DQ for writing.");
    } # end if
} # end sub

# Module return value:
1;
__END__

=head1 NAME

Text::BSV::BsvWriter - generates BSV data from a list of field names and a
list of references to hashes that encapsulate records.  Provides a method
for returning the BSV data as a reference to an array of strings, and a
method for writing the BSV data to a file.

=head1 SYNOPSIS

  use Text::BSV::BsvWriter;
  use Text::BSV::Exception;

  # Create a Text::BSV::BsvWriter instance:
  my $bsv_writer;

  eval {
      $bsv_writer = Text::BSV::BsvWriter->new($field_names, $records);
  };

  if ($EVAL_ERROR) {
      if ($EVAL_ERROR->get_type()
        == $Text::BSV::Exception::ILLEGAL_ARGUMENT) {
          say STDERR "Bad arguments passed to the Text::BSV::BsvWriter "
            . "constructor:  " . $EVAL_ERROR->get_message();
          exit(1);
      }
      else {
          say STDERR "An unknown exception occurred:  "
            . $EVAL_ERROR->get_message();
          exit(1);
      } # end if
  } # end if

  # Write the BSV data generated by the constructor to a file:
  eval {
      $bsv_writer->write_to_file($file_path);
  };

  if ($EVAL_ERROR) {
      say STDERR $EVAL_ERROR->get_message();
      exit(1);
  } # end if

  # Get the BSV data rows in case you want to do something with them other
  # than writing them to a file:
  my $rows = $bsv_writer->get_rows();

=head1 DESCRIPTION

This module defines a class for generating BSV data from a list of field
names and a list of references to hashes that encapsulate records.  The
class provides a method for returning the BSV data as a reference to an
array of strings, and a method for writing the BSV data to a file.

For a complete specification of the BSV (Bar-Separated Values) format, see
F<bsv_format.txt>.

In addition to the class-name argument, which is passed in automatically
when you use the C<Text::BSV::BsvWriter-E<gt>new()> syntax, the
constructor takes a reference to an array of field names and a reference to
an array of references to hashes encapsulating the records.

The constructor returns a reference to a Text::BSV::BsvWriter object,
which is implemented internally as a hash.  All functionality is exposed
through methods.

  NOTE:  This module uses the Text::BSV::Exception module for error
  handling.  When an error occurs during the execution of a method
  (including the constructor), the method creates a new
  Text::BSV::Exception object of the appropriate type and then passes
  it to "die".  When you call the constructor or a method documented
  to throw an exception, do so within an "eval" statement and then
  query $EVAL_ERROR ($@) to catch any exceptions that occurred.  For
  more information, see the documentation for Text::BSV::Exception.

=head1 PREREQUISITES

This module requires Perl 5 (version 5.10.1 or later), the
Text::BSV::BsvParsing module, and the Text::BSV::Exception module.

=head1 METHODS

=over

=item Text::BSV::BsvWriter->new($field_names, $records);

This is the constructor.  If the field names are not unique or do not
match the keys in each hash, or if there are not at least two fields
and at least one record, the constructor throws an exception of type
$Text::BSV::Exception::ILLEGAL_ARGUMENT.

=item $bsv_writer->get_rows();

Returns a reference to an array of rows (including the header row)
that compose the BSV data.  The rows returned by this method do not
have any end-of-line characters added.

=item $bsv_writer->write_to_file();

Takes the path to a BSV file (which will be overwritten if it already
exists) and writes the BSV data generated by the constructor to the
file.  If the file cannot be written to, the method throws an
exception of type $Text::BSV::Exception::IO_ERROR.

=back

=head1 AUTHOR

Benjamin Fitch, <blernflerkl@yahoo.com>

=head1 COPYRIGHT AND LICENSE

Copyright 2010 by Benjamin Fitch.

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


lib/Text/BSV/BsvParsing.pm
####
# BsvParsing.pm:  A Perl module containing helper functions for parsing and
# generating BSV data.
#
#   NOTE:  For a complete specification of the BSV (Bar-Separated Values)
#   format, see "bsv_format.txt".
#
####
#
# Copyright 2010 by Benjamin Fitch.
#
# This library is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.
####
package Text::BSV::BsvParsing;

use 5.010001;
use strict;
use warnings;
use utf8;

use English "-no_match_vars";
use Hash::Util ("lock_keys");
use List::Util ("first", "max", "min", "sum");
use Scalar::Util ("looks_like_number");
use Exporter;

use Text::BSV::Exception;

# Version:
our $VERSION = '1.03';

# Specify default exports:
our @ISA = ("Exporter");
our @EXPORT = (
  "get_field_delimiter",
  "parse_header_row",
  "parse_row",
  "generate_header_row",
  "generate_row",
  );

# Constants:
my $POUND     = "#";
my $SQ        = "'";
my $DQ        = "\"";
my $SEMICOLON = ";";
my $CR        = "\r";
my $LF        = "\n";
my $SPACE     = " ";
my $EMPTY     = "";
my $TRUE      = 1;
my $FALSE     = 0;

my $INCLUDE_TRAILING_EMPTY_FIELDS = -1;

####
# Exported helper functions for use by the other Text::BSV::* modules:
####

# The get_field_delimiter() function takes a string containing the header
# row and a string containing the first non-header row.  The function
# returns a string containing the field delimiter.
#
# If the field delimiter cannot be unambiguously determined, the function
# throws an exception of type $Text::BSV::Exception::INVALID_DATA_FORMAT.
sub get_field_delimiter {
    my $header_row = $_[0];
    my $first_nonheader_row = $_[1];
    my $field_delimiter;

    if (index($header_row, "|") > -1) {
        $field_delimiter = "|";
    }
    elsif (index($header_row, $DQ) > -1
      || index($first_nonheader_row, $DQ) > -1) {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::INVALID_DATA_FORMAT,
          "There is no vertical bar in the header row, but there is a "
          . "double quotation mark in the header row or in the first "
          . "non-header row.");
    }
    else {
        my @potential_delimiters;

        for my $delim (",", $SEMICOLON, "\t") {
            if (index($header_row, $delim) > -1) {
                push @potential_delimiters, $delim;
            } # end if
        } # next $delim

        given (scalar @potential_delimiters) {
            when (0) {
                die Text::BSV::Exception->new(
                  $Text::BSV::Exception::INVALID_DATA_FORMAT,
                  "The field delimiter in the BSV data cannot be "
                  . "unambiguously determined.");
            }
            when (1) {
                $field_delimiter = $potential_delimiters[0];
            }
            default {
                my $found_winner = $FALSE;

                for my $potential_delimiter (@potential_delimiters) {
                    if (num_delimiters($header_row, $potential_delimiter)
                      == num_delimiters(
                      $first_nonheader_row, $potential_delimiter)) {
                        if ($found_winner) {
                            die Text::BSV::Exception->new(
                              $Text::BSV::Exception::INVALID_DATA_FORMAT,
                              "The field delimiter in the BSV data "
                              . "cannot be unambiguously determined.");
                        }
                        else {
                            $found_winner = $TRUE;
                            $field_delimiter = $potential_delimiter;
                        } # end if
                    } # end if
                } # next $potential_delimiter

                unless ($found_winner) {
                    die Text::BSV::Exception->new(
                      $Text::BSV::Exception::INVALID_DATA_FORMAT,
                      "The field delimiter in the BSV data cannot be "
                          . "unambiguously determined.");
                } # end unless
            } # end when
        } # end given
    } # end if

    return $field_delimiter;
} # end sub

# The parse_header_row() function takes a string containing the header row
# (without end-of-line characters) and a string containing the field
# delimiter.  The function parses the header row, verifies that the field
# names are unique, and then returns a reference to an array of strings
# containing the field names.  If the field delimiter is the vertical bar,
# the function translates any escape sequences in the field names into the
# corresponding actual characters.
#
# If the header row contains any newline or carriage-return characters or
# the specified field delimiter is not supported by the BSV format, the
# function throws an exception of type
# $Text::BSV::Exception::ILLEGAL_ARGUMENT.  If the field names are not
# unique, or if the header row contains a double quotation mark but the
# field delimiter is not the vertical bar, the function throws an exception
# of type $Text::BSV::Exception::INVALID_DATA_FORMAT.
sub parse_header_row {
    my $header_row = $_[0];
    my $field_delimiter = $_[1];
    my @field_names;

    if ($header_row =~ /[\r\n]/s) {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::ILLEGAL_ARGUMENT,
          "The header row passed to the parse_header_row() function "
          . "contains newline or carriage-return characters.");
    } # end if

    if ($field_delimiter eq "|") {
        my $bsv_data = $header_row;
        my @pieces;

        $bsv_data =~ s/\\\\/\n/gs;
        @pieces = split /(?<!\\)\|/s, $header_row,
          $INCLUDE_TRAILING_EMPTY_FIELDS;

        for my $piece (@pieces) {
            my $field_name = $piece;

            $field_name =~ s/\n/\\\\/gs;
            $field_name = translate_from_bsv($field_name);

            if ($field_name ~~ @field_names) {
                die Text::BSV::Exception->new(
                  $Text::BSV::Exception::INVALID_DATA_FORMAT,
                  "Duplicate field names.");
            } # end if

            push @field_names, $field_name;
        } # next $piece
    }
    elsif (index($header_row, $DQ) > -1) {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::INVALID_DATA_FORMAT,
          "The field delimiter in the BSV data is not the vertical bar, "
          . "but there is a double quotation mark in the header row.");
    }
    elsif ($field_delimiter eq ","
      || $field_delimiter eq $SEMICOLON
      || $field_delimiter eq "\t") {
        my @pieces = split /$field_delimiter/s, $header_row,
          $INCLUDE_TRAILING_EMPTY_FIELDS;

        for my $piece (@pieces) {
            if ($piece ~~ @field_names) {
                die Text::BSV::Exception->new(
                  $Text::BSV::Exception::INVALID_DATA_FORMAT,
                  "Duplicate field names.");
            } # end if

            push @field_names, $piece;
        } # next $piece
    }
    else {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::ILLEGAL_ARGUMENT,
          "The field delimiter passed to the parse_header_row() function "
          . "is not supported by the BSV format.");
    } # end if

    return \@field_names;
} # end sub

# The parse_row() function takes a string containing a non-header row
# (without end-of-line characters), a string containing the field delimiter,
# and a reference to an array of strings containing the field names.  The
# function parses the row, verifies that the number of field values is
# correct, and then returns a reference to a hash in which the keys are the
# field names and the values are the record's field values.  If the field
# delimiter is the vertical bar, the function translates any escape
# sequences in the field values into the appropriate actual characters.
#
# The function assumes that the field names passed in are already translated
# from BSV format into ordinary text.
#
# If the row contains any newline or carriage-return characters or the
# specified field delimiter is not supported by the BSV format, the
# function throws an exception of type
# $Text::BSV::Exception::ILLEGAL_ARGUMENT.  If the row does not contain the
# number of fields matching the number of field names passed in, or if the
# row contains a double quotation mark but the field delimiter is not the
# vertical bar, the function throws an exception of type
# $Text::BSV::Exception::INVALID_DATA_FORMAT.
sub parse_row {
    my $row = $_[0];
    my $field_delimiter = $_[1];
    my $field_names = $_[2];
    my $record = {};

    if ($row =~ /[\r\n]/s) {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::ILLEGAL_ARGUMENT,
          "The row passed to the parse_row() function contains "
          . "newline or carriage-return characters.");
    } # end if

    if ($field_delimiter eq "|") {
        my $bsv_data = $row;
        my @pieces;

        $bsv_data =~ s/\\\\/\n/gs;
        @pieces = split /(?<!\\)\|/s, $row,
          $INCLUDE_TRAILING_EMPTY_FIELDS;

        unless (scalar(@pieces) == scalar(@{ $field_names })) {
            die Text::BSV::Exception->new(
              $Text::BSV::Exception::INVALID_DATA_FORMAT,
              "The number of fields in a row passed to the "
              . "parse_row() function does not match the number of "
              . "field names passed in.");
        } # end unless

        for my $dex (0..$#pieces) {
            my $field_value = $pieces[$dex];

            $field_value =~ s/\n/\\\\/gs;
            $field_value = translate_from_bsv($field_value);
            $record->{$field_names->[$dex]} = $field_value;
        } # next $piece
    }
    elsif (index($row, $DQ) > -1) {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::INVALID_DATA_FORMAT,
          "The field delimiter in the BSV data is not the vertical bar, "
          . "but there is a double quotation mark in at least one "
          . "of the rows.");
    }
    elsif ($field_delimiter eq ","
      || $field_delimiter eq $SEMICOLON
      || $field_delimiter eq "\t") {
        my @pieces = split /$field_delimiter/s, $row,
          $INCLUDE_TRAILING_EMPTY_FIELDS;

        unless (scalar(@pieces) == scalar(@{ $field_names })) {
            die Text::BSV::Exception->new(
              $Text::BSV::Exception::INVALID_DATA_FORMAT,
              "The number of fields in a row passed to the "
              . "parse_row() function does not match the number of "
              . "field names passed in.");
        } # end unless

        for my $dex (0..$#pieces) {
            $record->{$field_names->[$dex]} = $pieces[$dex];
        } # next $dex
    }
    else {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::ILLEGAL_ARGUMENT,
          "The field delimiter passed to the parse_row() function "
          . "is not supported by the BSV format.");
    } # end if

    return $record
} # end sub

# The generate_header_row() function takes a reference to an array of
# strings containing field names, and returns a string containing a BSV
# header row (with no end-of-line characters added).
#
# If the field names passed in are not unique, the function throws an
# exception of type $Text::BSV::Exception::ILLEGAL_ARGUMENT.
sub generate_header_row {
    my $field_names = $_[0];
    my @validated_field_names = ();
    my $header_row = $EMPTY;

    for my $dex (0..$#{ $field_names }) {
        if ($field_names->[$dex] ~~ @validated_field_names) {
            die Text::BSV::Exception->new(
              $Text::BSV::Exception::ILLEGAL_ARGUMENT,
              "Field names are not unique.");
        } # end if

        push @validated_field_names, $field_names->[$dex];

        if ($dex > 0) {
            $header_row .= "|";
        } # end if

        $header_row .= translate_to_bsv($field_names->[$dex]);
    } # next $dex

    return $header_row;
} # end sub

# The generate_row() function takes a reference to a hash that encapsulates
# a BSV record and a reference to an array of strings containing the field
# names.  Using the array of field names to determine the order of the
# fields, the function generates and returns a string containing a BSV
# non-header row (with no end-of-line characters added).
#
# If the number and names of the fields in the passed-in hash do not match
# the passed-in array of field names, the function throws an exception of
# type $Text::BSV::Exception::ILLEGAL_ARGUMENT.
sub generate_row {
    my $record = $_[0];
    my $field_names = $_[1];
    my $row = $EMPTY;

    unless (scalar(keys %{ $record }) == scalar(@{ $field_names })) {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::ILLEGAL_ARGUMENT,
          "The number of field names passed to the generate_row() function "
          . "does not match the number field values in the record.");
    } # end unless

    for my $dex (0..$#{ $field_names }) {
        unless (exists $record->{$field_names->[$dex]}) {
            die Text::BSV::Exception->new(
              $Text::BSV::Exception::ILLEGAL_ARGUMENT,
              "The record passed to the generate_row() function does not "
              . "match the list of field names passed in.");
        } # end unless

        if ($dex > 0) {
            $row .= "|";
        } # end if

        $row .= translate_to_bsv($record->{$field_names->[$dex]});
    } # next $dex

    return $row;
} # end sub

####
# Private helper functions for use by this module:
####

# The private num_delimiters() function takes a string to be searched and a
# string containing a delimiter.  The function returns the number of
# occurrences of the delimiter within the string to be searched.
sub num_delimiters {
    my $str = $_[0];
    my $delim = $_[1];
    my $result = 0;

    for my $dex (0..(length($str) - 1)) {
        if (substr($str, $dex, 1) eq $delim) {
            $result++;
        } # end if
    } # next $dex

    return $result;
} # end sub

# The private translate_from_bsv() function takes a string containing a BSV
# field name or value and returns a string containing the translated version
# in which escape sequences have been replaced with the correct
# corresponding characters.
#
# If the field name or value contains any instances of invalid backslash
# usage, the function throws an exception of type
# $Text::BSV::Exception::INVALID_DATA_FORMAT.  If the field name or value
# contains any newline or carriage-return characters, the function throws an
# exception of type $Text::BSV::Exception::ILLEGAL_ARGUMENT.
sub translate_from_bsv {
    my $bsv_str = $_[0];
    my $normal_str = $bsv_str;

    if ($bsv_str =~ /[\r\n]/s) {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::ILLEGAL_ARGUMENT,
          "A BSV field name or value cannot contain newline or "
          . "carriage-return characters.");
    } # end if

    $normal_str =~ s/\\\\/\n/gs;

    if ($normal_str =~ /\\[^|n]/s) {
        die Text::BSV::Exception->new(
          $Text::BSV::Exception::INVALID_DATA_FORMAT,
          "Invalid backslash usage in BSV data.");
    } # end if

    $normal_str =~ s/\n/\\z/gs;
    $normal_str =~ s/\\\|/|/gs;
    $normal_str =~ s/\\n/\n/gs;
    $normal_str =~ s/\\z/\\/gs;

    return $normal_str;
} # end sub

# The private translate_to_bsv() function takes a string containing a
# field name or value and returns a string containing the BSV version,
# in which carriage returns have been stripped and then backslashes,
# vertical bars, and newline characters have been escaped.
sub translate_to_bsv {
    my $bsv_str = $_[0];

    $bsv_str =~ s/\r/$EMPTY/gs;
    $bsv_str =~ s/\\/\\\\/gs;
    $bsv_str =~ s/\|/\\|/gs;
    $bsv_str =~ s/\n/\\n/gs;

    return $bsv_str;
} # end sub

# Module return value:
1;


lib/Text/BSV/Exception.pm
####
# Exception.pm:  A Perl module defining a package for creating a
# Text::BSV::Exception object, which can be thrown using the Perl "die"
# function and caught using "eval".
#
# In addition to the class-name argument, which is passed in automatically
# when you use the Text::BSV::Exception->new() syntax, the constructor takes
# an exception type (one of the contants defined within the package), a
# message string, and an optional hash reference that points to arbitrary
# structured data pertaining to the exception.
#
# The constructor returns a reference to a Text::BSV::Exception object,
# which is implemented internally as a hash.
#
# The Text::BSV::Exception package provides the following public
# accessor methods:
#
#     get_type()
#     get_message()
#     get_data()
#
# If there is no structured data, the get_data() method returns a reference
# to an empty hash.
#
####
#
# Copyright 2010 by Benjamin Fitch.
#
# This library is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.
####
package Text::BSV::Exception;

use 5.010001;
use strict;
use warnings;
use utf8;

use English "-no_match_vars";
use Hash::Util ("lock_keys");
use List::Util ("first", "max", "min", "sum");
use Scalar::Util ("looks_like_number");

# Version:
our $VERSION = '1.03';

# Exception types:
our $DIRECTORY_NOT_FOUND   = 0;
our $FILE_NOT_FOUND        = 1;
our $GENERAL               = 2;
our $ILLEGAL_ARGUMENT      = 3;
our $INVALID_CHARACTER     = 4;
our $INVALID_DATA_FORMAT   = 5;
our $INVALID_REGEX         = 6;
our $IO_ERROR              = 7;
our $NULL_POINTER          = 8;
our $UNSUPPORTED_OPERATION = 9;

# Reference types:
my $SCALAR_TYPE = ref \(my $foo);
my $ARRAY_TYPE  = ref [];
my $HASH_TYPE   = ref {};
my $CODE_TYPE   = ref sub { };
my $REGEX_TYPE  = ref qr//;

# General constants:
my $POUND     = "#";
my $SQ        = "'";
my $DQ        = "\"";
my $SEMICOLON = ";";
my $CR        = "\r";
my $LF        = "\n";
my $SPACE     = " ";
my $EMPTY     = "";
my $TRUE      = 1;
my $FALSE     = 0;

# Constructor:
sub new {
    my ($class, @args) = @_;
    my %exception = ("_type" => $args[0], "_message" => $args[1]);

    unless (defined($exception{"_type"})
      && looks_like_number($exception{"_type"})
      && $exception{"_type"} == int($exception{"_type"})
      && $exception{"_type"} >= 0 && $exception{"_type"} <= 9) {
        $exception{"_type"} = $GENERAL;
    } # end unless

    unless (defined $exception{"_message"}) {
        $exception{"_message"} = $EMPTY;
    } # end unless

    $exception{"_data"} = ref($args[2]) eq $HASH_TYPE
      ? $args[2] : {};

    # Bless the hash into the class:
    bless \%exception, $class;

    # Restrict the hash keys:
    lock_keys(%exception, "_type", "_message", "_data");

    # Return the object:
    return \%exception;
} # end constructor

# Methods:
sub get_type {
    return $_[0]->{"_type"};
} # end sub

sub get_message {
    return $_[0]->{"_message"};
} # end sub

sub get_data {
    return $_[0]->{"_data"};
} # end sub

# Module return value:
1;
__END__

=head1 NAME

Text::BSV::Exception - create an object that can be thrown using the Perl
"die" function (which can accept a reference instead of a string) and caught
using "eval", which stores the object in $EVAL_ERROR ($@).

=head1 SYNOPSIS

  use Text::BSV::Exception;
  my $exception = Text::BSV::Exception->new($Text::BSV::Exception::GENERAL,
    "Couldn't do stuff because of the thing.");
  my $fancy_exception = Text::BSV::Exception->new(
    $Text::BSV::Exception::NULL_POINTER,
    "Couldn't do stuff because the thing was pointing to nothing.",
    {"IsFatal" => $TRUE, "Time" => "17:47:18"});
  my $exception_type = $exception->get_type();
  my $message_string = $exception->get_message();
  my $hash_ref = $fancy_exception->get_data();

=head1 DESCRIPTION

This module defines a package for creating a Text::BSV::Exception object,
which can be thrown using the Perl "die" function and caught using "eval".

In addition to the class-name argument, which is passed in automatically
when you use the C<Text::BSV::Exception-E<gt>new()> syntax, the constructor
takes an exception type, a message string, and an optional hash reference
that points to arbitrary structured data pertaining to the exception.

The exception type must be one of the following constants defined in the
Text::BSV::Exception package:

    $DIRECTORY_NOT_FOUND
    $FILE_NOT_FOUND
    $GENERAL
    $ILLEGAL_ARGUMENT
    $INVALID_CHARACTER
    $INVALID_DATA_FORMAT
    $INVALID_REGEX
    $IO_ERROR
    $NULL_POINTER
    $UNSUPPORTED_OPERATION

The constructor returns a reference to a Text::BSV::Exception object,
which is implemented internally as a hash.  All functionality is exposed
through methods.

=head1 PREREQUISITES

This module requires Perl 5, version 5.10.1 or later.

=head1 METHODS

=over

=item Text::BSV::Exception->new($exception_type, $message,
  $optional_hash_ref);

This is the constructor.

=item $exception->get_type();

This is the accessor method for retrieving the exception type, which
must be one of the following constants defined in the
Text::BSV::Exception package:

    $DIRECTORY_NOT_FOUND
    $FILE_NOT_FOUND
    $GENERAL
    $ILLEGAL_ARGUMENT
    $INVALID_CHARACTER
    $INVALID_DATA_FORMAT
    $INVALID_REGEX
    $IO_ERROR
    $NULL_POINTER
    $UNSUPPORTED_OPERATION

=item $exception->get_message();

This is the accessor method for retrieving the message.

=item $exception->get_data();

This is the accessor method for retrieving any structured data provided
with the exception.  If there is no structured data, the C<get_data()>
method returns a reference to an empty hash.

=back

=head1 AUTHOR

Benjamin Fitch, <blernflerkl@yahoo.com>

=head1 COPYRIGHT AND LICENSE

Copyright 2010 by Benjamin Fitch.

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
